File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
DIRS = [(0, 1), (0, -1), (1, 0), (-1,0)]

T = int(input())

for tc in range(1, T+1):
    N, M , K = map(int,input().split())
    cabbage_field = [[0] * M for _ in range(N)]
    for _ in range(K):
        a, b = map(int,input().split())
        cabbage_field[a][b] = 1

    visited = [[False] * M for _ in range(N)]

    stk = []
    cnt = 0
    for y in range(N):
        for x in range(M):
            if cabbage_field[y][x] == 1:
                if visited[y][x]:
                    continue
                stk.append((y, x))
                cnt += 1
                while stk:
                    sy, sx = stk.pop()
                    visited[sy][sx] = True

                    for dy, dx in DIRS:
                        ny, nx = dy + sy, dx + sx
                        if ny < 0 or ny >= N or nx < 0 or nx >= M:
                            continue
                        if visited[ny][nx] or cabbage_field[ny][nx] == 0:
                            continue
                        stk.append((ny, nx))
    
    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int DX[4] = { 0, 0, 1, -1 };
int DY[4] = { 1, -1, 0, 0 };

struct Point {
    int y;
    int x;
};

int main() {
    int T;
    cin >> T;

    for (int tc = 0; tc < T; tc++) {
        int M, N, K;
        int cnt = 0;
        cin >> M >> N >> K;

        vector<vector<int>> matrix(N, vector<int>(M, 0));

        for (int i = 0; i < K; i++) {
            int cabx, caby;
            cin >> cabx >> caby;
            matrix[caby][cabx] = 1;
        }

        vector<vector<bool>> visited(N, vector<bool>(M, 0));

        queue<Point> Q;

        for (int y = 0; y < N; y++) {
            for (int x = 0; x < M; x++) {
                if (!matrix[y][x]) {
                    continue;
                }
                if (visited[y][x]) {
                    continue;
                }

                cnt++;

                Q.push({ y, x });
                visited[y][x] = 1;

                while (!Q.empty()) {
                    Point P = Q.front();
                    Q.pop();

                    for (int i = 0; i < 4; i++) {
                        int dy = P.y + DY[i];
                        int dx = P.x + DX[i];

                        if (dy < 0 || dy >= N || dx < 0 || dx >= M) {
                            continue;
                        }
                        if (!matrix[dy][dx] || visited[dy][dx]) {
                            continue;
                        }

                        visited[dy][dx] = 1;
                        Q.push({ dy, dx });
                    }
                }
            }
        }
        cout << cnt << endl;

    }

    return 0;
}
